### PR TITLE
systemctl.c, cgroup.h, cgroup.c, dbus-unit.c: Let systemctl print IO accounting information

### LINKED ISSUE #9768: Show total IO in systemctl
**systemd version the issue has been seen with**
 239

**Used distribution**

Debian unstable in systemd-nspawn on Ubuntu 18.04 LTS

**Expected behaviour you didn't see**

With `IOAccounting=yes` and existence of `io.stat` in a directory corresponding to a unit,
`systemctl status someunit.service` etc. should show the information stored in `io.stat`

**Unexpected behaviour you saw**

Systemd nor systemctl reads information from `io.stat`. I run

```
fgrep -n "io.stat" `find -type f -print`
```

on the latest source of systemd, and found that `io.stat` is only referred in `src/cgtop/cgtop.c:334`,
so there is **absolutely no hope** for `systemctl` to show information in `io.stat`.

In addition no option to `systemd-cgtop` lets `systemd-cgtop` to print information from `io.stat`.
Then, what is the intention of `IOAccouting=yes` in the unified cgroup hierarchy??
I cannot understand what is meant by systemd.resource-control manpage.
What is the expected behavior **?????????????????????????**

I did the following:

```
root@gtx1080ti-unstable:/etc/systemd/system# systemctl show user-0.slice | fgrep Accounting
CPUAccounting=yes
IOAccounting=yes
BlockIOAccounting=no
MemoryAccounting=yes
TasksAccounting=yes
IPAccounting=yes

root@gtx1080ti-unstable:/etc/systemd/system# cat /sys/fs/cgroup/user.slice/user-0.slice/io.stat 
8:0 rbytes=1425408 wbytes=20480 rios=36 wios=1

root@gtx1080ti-unstable:/etc/systemd/system# systemctl status -l  user-0.slice
● user-0.slice - User Slice of UID 0
   Loaded: loaded
  Drop-In: /lib/systemd/system/user-.slice.d
           └─10-defaults.conf
           /etc/systemd/system/user-.slice.d
           └─accounting.conf
   Active: active since Wed 2018-08-01 16:48:52 JST; 23min ago
    Tasks: 6 (limit: 10813)
   Memory: 11.5M
      CPU: 1.578s
   CGroup: /user.slice/user-0.slice
           ├─session-345.scope
           │ ├─1866 sshd: root@pts/0
           │ ├─1879 -bash
           │ ├─1973 systemctl status -l user-0.slice
           │ └─1974 pager
           └─user@0.service
             └─init.scope
               ├─1869 /lib/systemd/systemd --user
               └─1870 (sd-pam)
```

**Steps to reproduce the problem**

See above.

### PR DESCRIPTION
**Pull request by ghost**
This PR lets systemctl print IO accounting information and closes #9768

### CODE REVIEW

**Comment by lucaswerkmeister** [comment]
A possible addition to this would be to also include IO statistics in the log message for a unit’s resource use (see `unit_log_resources()` in `unit.c`), like CPU time and IP traffic.

**Comment by ryutaroh-matsumoto** [comment]
https://github.com/lucaswerkmeister

**Comment by ryutaroh-matsumoto** [comment]
@lucaswerkmeister thank you for your comment. I agree and if the proposed change is judged as acceptable then I will try to add IO statistics to unit_log_resources and systemd-run --wait

File: src/core/cgroup.c [hunk 1]
```diff
@@ -2584,6 +2584,73 @@ int unit_get_ip_accounting(
         return r;
 }
 
+/* The below is borrowed from cgtop/cgtop.c */
```
Comment (keszybz) [inline src/core/cgroup.c:2587]: You should not just take a long block of code like this and copy it from one file to another. Instead, this needs to be turned into a function somewhere in `src/shared` and called from both locations.

File: src/core/cgroup.c [hunk 2]
```diff
@@ -2584,6 +2584,73 @@ int unit_get_ip_accounting(
...
+        if (unit_has_root_cgroup(u))
+                return -ENODATA;
+
+        r = cg_get_path("io", u->cgroup_path, "io.stat", &p);
+        if (r < 0)
+                return -ENODATA;
+
+        f = fopen(p, "re");
+        if (!f)
+                return -ENODATA;
```
Comment (keszybz) [inline src/core/cgroup.c:2619]: Unlike the original, this returns a fake error value. The error can be anything like -EPERM, -ENOMEM, -ENOFILE, etc. Do the same as original, and translate -ENOENT into -ENODATA, but propagate any other error.

File: src/core/cgroup.c [hunk 3]
```diff
@@ -2584,6 +2584,73 @@ int unit_get_ip_accounting(
...
+                        l += strspn(l, WHITESPACE);
+                }
+        }
+
+        if (metric == CGROUP_IO_INGRESS_BYTES)
+                *ret = rd;
+        else if (metric == CGROUP_IO_EGRESS_BYTES)
+                *ret = wr;
+        else
+                assert(metric == CGROUP_IO_INGRESS_BYTES || metric == CGROUP_IO_EGRESS_BYTES);
```
Comment (keszybz) [inline src/core/cgroup.c:2649]: No need to repeat the condition here. Use assert_not_reached() instead.

**Comment by ryutaroh-matsumoto** [comment]
Sorry, I don't have time to update this PR in near future, so I close this PR.

**Pull request closed without merging**